gui/src/hw.rs

```diff
@@ -341,60 +468,156 @@ impl HardwareWallets {
         }
     }
 
-    pub fn refresh(&self) -> iced::Subscription<HardwareWalletMessage> {
+    pub fn reset_refresh(&mut self) {
+        // In order to reset the subscription state we need to change the
+        // subscription id
+        self.refresh_index = self.refresh_index.wrapping_add(1);
+    }
+
+    pub fn refresh(&self, taproot: bool) -> iced::Subscription<HardwareWalletMessage> {
+        let id = format!(
+            "refresh-{}-{}-{}",
+            self.network, self.refresh_index, taproot
+        );
         iced::subscription::unfold(
-            format!("refresh-{}", self.network),
+            id,
             State {
                 network: self.network,
                 keys_aliases: self.aliases.clone(),
                 wallet: self.wallet.clone(),
                 connected_supported_hws: Vec::new(),
+                hws_upgrade: Vec::new(),
                 api: None,
                 datadir_path: self.datadir_path.clone(),
+                taproot,
+                hws: Vec::new(),
+                still: Vec::new(),
+                still_upgrade: Vec::new(),
+                sleep: false,
             },
             refresh,
         )
     }
 }
 
-struct State {
+pub struct State {
     network: Network,
     keys_aliases: HashMap<Fingerprint, String>,
     wallet: Option<Arc<Wallet>>,
     connected_supported_hws: Vec<String>,
     api: Option<ledger::HidApi>,
     datadir_path: PathBuf,
+    taproot: bool,
+    hws: Vec<HardwareWallet>,
+    still: Vec<String>,
+    hws_upgrade: Vec<HardwareWallet>,
+    still_upgrade: Vec<String>,
+    sleep: bool,
 }
 
 async fn refresh(mut state: State) -> (HardwareWalletMessage, State) {
-    let api = if let Some(api) = &mut state.api {
-        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
-        if let Err(e) = api.refresh_devices() {
-            return (HardwareWalletMessage::Error(e.to_string()), state);
-        };
+    // do not sleep on first call
```

Comment:
We do not need the .sleep field, we can use `state.api.is_some()` to detect if it is the first call